samples/python_interop/cirq_samples/custom_gates.ipynb

```diff
@@ -0,0 +1,241 @@
+    "    program.append(cirq.measure(bob, key='result'))\n",
+    "    return program"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "54968cd9",
+   "metadata": {},
+   "source": [
+    "## 4. Inspect the Circuit with Custom Gates"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "8d8ce308",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "circuit = cirq.Circuit(simple_teleportation())\n",
+    "print('Circuit with custom gates:')\n",
+    "print(circuit)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "8c98e44a",
+   "metadata": {},
+   "source": [
+    "## 5. Full Decomposition\n",
+    "We can ask Cirq to decompose all gates to primitive operations."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "068d7430",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "decomposed_full = cirq.Circuit(cirq.decompose(circuit))\n",
+    "print('Fully decomposed:')\n",
+    "print(decomposed_full)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "c9e98407",
+   "metadata": {},
+   "source": [
+    "## 6. Selective Decomposition\n",
+    "\n",
+    "The `keep` predicate controls where recursive decomposition stops:\n",
+    "- If `keep(op)` returns True, the operation is left as-is.\n",
+    "- If it returns False, Cirq tries to expand the operation via its decomposition methods.\n",
+    "\n",
+    "In the example below, we decompose only the custom gates (`BellPairGate`, `MessagePrepGate`) by returning False for them and True for everything else."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "ac09e4b7",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def keep(op: cirq.Operation) -> bool:\n",
+    "    gate = getattr(op, 'gate', None)\n",
+    "    return not isinstance(gate, (BellPairGate, MessagePrepGate))\n",
+    "\n",
+    "decomposed_custom = cirq.Circuit(cirq.decompose(circuit, keep=keep))\n",
+    "print('Decomposed custom gates removed:')\n",
+    "print(decomposed_custom)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "6faa0c1e",
+   "metadata": {},
+   "source": [
+    "## 7. Simulation\n",
+    "\n",
+    "We run the teleportation circuit several times to confirm Bob receives (approximately) the same state that was prepared on the message qubit. Because we removed the final inverse rotation, Bob's measurement statistics should reflect |ψ> = Rx(0.7)|0>. Expected probabilities: P(0) ≈ cos^2(0.35) ≈ 0.88, P(1) ≈ sin^2(0.35) ≈ 0.12."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "1bdc4615",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import math\n",
+    "\n",
+    "sim = cirq.Simulator()\n",
+    "result = sim.run(cirq.Circuit(simple_teleportation()), repetitions=1000)\n",
+    "print(\"Histogram of teleported state measurements (expected ~88% zeros, ~12% ones):\")\n",
```

Comment:
If you print the expected probabilities maybe there's no need to have them as percentages in the hard-coded string. Also, maybe adding this calculation to the teleportation notebook would help.

Reply:
They don't need to be rewritten as hard-coded strings in the code. But I do like having the expected state stated in the markdown cell. Hard to describe why, but I feel like it is missing something without it.